### PR TITLE
SONARJAVA-1431 Extract rule metadatas in json file

### PR DESCRIPTION
**Pull request by benzonico**
This implied :
Change the way rules are created.
Change the way Sonar Way profile is created.
Read json files.
Map legacy keys to RSPEC keys in rules annotation requiring it

### CODE REVIEW

**Comment by SonarTech** [comment]
#### Analysis summary
SonarQube analysis reported 1 issue

#### 1 unreported issues

Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:

1. org.sonarsource.java:java-checks:src/main/java/org/sonar/java: MAJOR Add a 'package-info.java' file to document the 'java' package [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1228)

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. MAJOR org.sonarsource.java:java-checks:src/main/java/org/sonar/java: Add a 'package-info.java' file to document the 'java' package [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1228)

**Comment by m-g-sonar** [comment]
Commit https://github.com/SonarSource/sonar-java/pull/780/commits/eb3ad93be1a8312cb2dd3b4f0dda0bc5f7624636 only update Sqale sub characteristic, and does not handle rule parameters.

File: java-checks/src/main/java/org/sonar/java/checks/CycleBetweenPackagesCheck.java [hunk 1]
```diff
@@ -23,6 +23,7 @@
 import org.sonar.api.server.rule.RulesDefinition;
```
Comment (SonarTech) [inline java-checks/src/main/java/org/sonar/java/checks/CycleBetweenPackagesCheck.java]: MINOR Remove this unused import 'org.sonar.api.server.rule.RulesDefinition'. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AUselessImportCheck)

File: java-checks/src/main/java/org/sonar/java/checks/CycleBetweenPackagesCheck.java [hunk 2]
```diff
@@ -23,6 +23,7 @@
...
 import org.sonar.check.Priority;
```
Comment (SonarTech) [inline java-checks/src/main/java/org/sonar/java/checks/CycleBetweenPackagesCheck.java]: MINOR Remove this unused import 'org.sonar.check.Priority'. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AUselessImportCheck)

File: java-checks/src/main/java/org/sonar/java/checks/CycleBetweenPackagesCheck.java [hunk 3]
```diff
@@ -23,6 +23,7 @@
...
 import org.sonar.check.Rule;
+import org.sonar.java.RspecKey;
 import org.sonar.java.bytecode.visitor.BytecodeVisitor;
 import org.sonar.java.tag.Tag;
 import org.sonar.squidbridge.annotations.SqaleConstantRemediation;
```
Comment (SonarTech) [inline java-checks/src/main/java/org/sonar/java/checks/CycleBetweenPackagesCheck.java]: MINOR Remove this unused import 'org.sonar.squidbridge.annotations.SqaleConstantRemediation'. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AUselessImportCheck)

File: java-checks/src/main/java/org/sonar/java/checks/CycleBetweenPackagesCheck.java [hunk 4]
```diff
@@ -23,6 +23,7 @@
 import org.sonar.api.server.rule.RulesDefinition;
 import org.sonar.check.Priority;
 import org.sonar.check.Rule;
+import org.sonar.java.RspecKey;
 import org.sonar.java.bytecode.visitor.BytecodeVisitor;
 import org.sonar.java.tag.Tag;
```
Comment (SonarTech) [inline java-checks/src/main/java/org/sonar/java/checks/CycleBetweenPackagesCheck.java]: MINOR Remove this unused import 'org.sonar.java.tag.Tag'. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AUselessImportCheck)

**Comment by SonarTech** [comment]
SonarQube analysis reported 11 issues
* MINOR 11 minor

Watch the comments in this conversation to review them.

#### 7 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. MINOR [CycleBetweenPackagesCheck.java#L30](https://github.com/SonarSource/sonar-java/blob/db40937fee8907372204a6210044730e747c47ae/java-checks/src/main/java/org/sonar/java/checks/CycleBetweenPackagesCheck.java#L30): Remove this unused import 'org.sonar.squidbridge.annotations.SqaleSubCharacteristic'. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AUselessImportCheck)
1. MINOR [EqualsNotOverriddenInSubclassCheck.java#L23](https://github.com/SonarSource/sonar-java/blob/db40937fee8907372204a6210044730e747c47ae/java-checks/src/main/java/org/sonar/java/checks/EqualsNotOverriddenInSubclassCheck.java#L23): Remove this unused import 'org.sonar.api.server.rule.RulesDefinition'. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AUselessImportCheck)
1. MINOR [EqualsNotOverriddenInSubclassCheck.java#L24](https://github.com/SonarSource/sonar-java/blob/db40937fee8907372204a6210044730e747c47ae/java-checks/src/main/java/org/sonar/java/checks/EqualsNotOverriddenInSubclassCheck.java#L24): Remove this unused import 'org.sonar.check.Priority'. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AUselessImportCheck)
1. MINOR [EqualsNotOverriddenInSubclassCheck.java#L27](https://github.com/SonarSource/sonar-java/blob/db40937fee8907372204a6210044730e747c47ae/java-checks/src/main/java/org/sonar/java/checks/EqualsNotOverriddenInSubclassCheck.java#L27): Remove this unused import 'org.sonar.java.tag.Tag'. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AUselessImportCheck)
1. MINOR [EqualsNotOverriddenInSubclassCheck.java#L38](https://github.com/SonarSource/sonar-java/blob/db40937fee8907372204a6210044730e747c47ae/java-checks/src/main/java/org/sonar/java/checks/EqualsNotOverriddenInSubclassCheck.java#L38): Remove this unused import 'org.sonar.squidbridge.annotations.ActivatedByDefault'. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AUselessImportCheck)
1. MINOR [EqualsNotOverriddenInSubclassCheck.java#L39](https://github.com/SonarSource/sonar-java/blob/db40937fee8907372204a6210044730e747c47ae/java-checks/src/main/java/org/sonar/java/checks/EqualsNotOverriddenInSubclassCheck.java#L39): Remove this unused import 'org.sonar.squidbridge.annotations.SqaleConstantRemediation'. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AUselessImportCheck)
1. MINOR [EqualsNotOverriddenInSubclassCheck.java#L40](https://github.com/SonarSource/sonar-java/blob/db40937fee8907372204a6210044730e747c47ae/java-checks/src/main/java/org/sonar/java/checks/EqualsNotOverriddenInSubclassCheck.java#L40): Remove this unused import 'org.sonar.squidbridge.annotations.SqaleSubCharacteristic'. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AUselessImportCheck)

**Comment by SonarTech** [comment]
SonarQube analysis reported 11 issues
* MINOR 11 minor

Watch the comments in this conversation to review them.

#### 7 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. MINOR [CycleBetweenPackagesCheck.java#L30](https://github.com/SonarSource/sonar-java/blob/3369b811a869242f02d019ce1fbbfb5479c4e753/java-checks/src/main/java/org/sonar/java/checks/CycleBetweenPackagesCheck.java#L30): Remove this unused import 'org.sonar.squidbridge.annotations.SqaleSubCharacteristic'. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AUselessImportCheck)
1. MINOR [EqualsNotOverriddenInSubclassCheck.java#L23](https://github.com/SonarSource/sonar-java/blob/3369b811a869242f02d019ce1fbbfb5479c4e753/java-checks/src/main/java/org/sonar/java/checks/EqualsNotOverriddenInSubclassCheck.java#L23): Remove this unused import 'org.sonar.api.server.rule.RulesDefinition'. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AUselessImportCheck)
1. MINOR [EqualsNotOverriddenInSubclassCheck.java#L24](https://github.com/SonarSource/sonar-java/blob/3369b811a869242f02d019ce1fbbfb5479c4e753/java-checks/src/main/java/org/sonar/java/checks/EqualsNotOverriddenInSubclassCheck.java#L24): Remove this unused import 'org.sonar.check.Priority'. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AUselessImportCheck)
1. MINOR [EqualsNotOverriddenInSubclassCheck.java#L27](https://github.com/SonarSource/sonar-java/blob/3369b811a869242f02d019ce1fbbfb5479c4e753/java-checks/src/main/java/org/sonar/java/checks/EqualsNotOverriddenInSubclassCheck.java#L27): Remove this unused import 'org.sonar.java.tag.Tag'. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AUselessImportCheck)
1. MINOR [EqualsNotOverriddenInSubclassCheck.java#L38](https://github.com/SonarSource/sonar-java/blob/3369b811a869242f02d019ce1fbbfb5479c4e753/java-checks/src/main/java/org/sonar/java/checks/EqualsNotOverriddenInSubclassCheck.java#L38): Remove this unused import 'org.sonar.squidbridge.annotations.ActivatedByDefault'. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AUselessImportCheck)
1. MINOR [EqualsNotOverriddenInSubclassCheck.java#L39](https://github.com/SonarSource/sonar-java/blob/3369b811a869242f02d019ce1fbbfb5479c4e753/java-checks/src/main/java/org/sonar/java/checks/EqualsNotOverriddenInSubclassCheck.java#L39): Remove this unused import 'org.sonar.squidbridge.annotations.SqaleConstantRemediation'. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AUselessImportCheck)
1. MINOR [EqualsNotOverriddenInSubclassCheck.java#L40](https://github.com/SonarSource/sonar-java/blob/3369b811a869242f02d019ce1fbbfb5479c4e753/java-checks/src/main/java/org/sonar/java/checks/EqualsNotOverriddenInSubclassCheck.java#L40): Remove this unused import 'org.sonar.squidbridge.annotations.SqaleSubCharacteristic'. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AUselessImportCheck)

File: java-checks-testkit/src/main/java/org/sonar/java/checks/verifier/CheckVerifier.java [hunk 5]
```diff
@@ -176,14 +185,33 @@ private void assertMultipleIssue(Set<AnalyzerMessage> issues) throws AssertionEr
     }
   }
 
-  private static void validateIssue(Multimap<Integer, Map<IssueAttribute, String>> expected, List<Integer> unexpectedLines, AnalyzerMessage issue) {
+  private boolean isLinear(AnalyzerMessage issue) {
```
Comment (SonarTech) [inline java-checks-testkit/src/main/java/org/sonar/java/checks/verifier/CheckVerifier.java]: MAJOR Make "isLinear" a "static" method. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS2325)

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue
* MAJOR 1 major

Watch the comments in this conversation to review them.

**Pull request merged**